File: 프로그래머스/Level 2/호텔 대실.py
# https://school.programmers.co.kr/learn/courses/30/lessons/155651
def transMinute(strTime):
    h,m = strTime.split(":")
    return int(h)*60+int(m)

def solution(book_time):
    answer = 0
    bookTime = []
    roomArr = []
    
    # 편의로 분단위로 바꿔줌
    for book in book_time:
        bookTime.append([transMinute(book[0]), transMinute(book[1]) + 10]) # 청소시간도 같이
    
    bookTime.sort() # 시작 시간이 빠른 순으로
    #print(bookTime)
    
    for book in bookTime:
        if not roomArr:
            roomArr.append(book)
            continue
        
        # roomArr에 넣을 수 있는지 체크
        for i in range(len(roomArr)):
            # 이번 예약시간이 현재 룸의 끝나는 시간과 같다면
            if book[0] >= roomArr[i][1]:
                roomArr[i][0] = book[0] # 교체
                roomArr[i][1] = book[1]
                break
        else:
            roomArr.append(book)
    #print(roomArr)
    answer = len(roomArr)
    return answer

File: 프로그래머스/Level 2/호텔 대실.swift
// https://school.programmers.co.kr/learn/courses/30/lessons/155651
import Foundation

func solution(_ book_time:[[String]]) -> Int {
    var answer:Int = 0
    var bookTime:[[Int]] = []
    var roomArr:[[Int]] = []
    
    func getTotalMinu(_ minu: String) -> Int {
        let splitedMin = minu.components(separatedBy: ":")
        
        return Int(splitedMin[0])!*60+Int(splitedMin[1])!
    }
    
    for book in book_time {
        let start = getTotalMinu(book[0])
        let end = getTotalMinu(book[1])
        bookTime.append([start, end])
    }
    
    bookTime.sort { $0[0] < $1[0] }

    
    for book in bookTime {
        if roomArr.isEmpty {
            roomArr.append(book)
        } else {
            var check:Bool = false
            for i in 0..<roomArr.count {
                if roomArr[i][1]+10 <= book[0] {
                    roomArr[i] = book
                    check = true
                    break
                }
            }
            if !check {
                roomArr.append(book)
            }
        }
    }
    
    answer = roomArr.count
    
    return answer
}